feat: add C implementation for `stats/base/dists/frechet/logpdf`

iraandrushko opened the pull request:
Resolves [#3608](https://github.com/stdlib-js/stdlib/issues/3608)

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds C implementation for @stdlib/stats/base/dists/frechet/logpdf along with relevant tests, benchmarks and examples.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3608  

Parent issue: #3564 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

iraandrushko committed:
feat: add C implementation for stats/base/dists/frechet/logpdf

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented:
/stdlib merge

iraandrushko committed:
feat: add C implementation for stats/base/dists/frechet/logpdf: fix markdown-examples-javascript files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/src/main.c:
```diff
@@ -0,0 +1,56 @@
+#include "stdlib/constants/float64/nan.h"
+
+/**
+* Returns the logpdf for a Fréchet distribution with shape `alpha`, scale `s`, and location `m` at a value `x`.
```
```suggestion
* Evaluates the logarithm of the probability density function (PDF) for a Fréchet distribution with shape `alpha`, scale `s`, and location `m` at a value `x`.
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/src/main.c:
```diff
@@ -0,0 +1,56 @@
+/**
+* Returns the logpdf for a Fréchet distribution with shape `alpha`, scale `s`, and location `m` at a value `x`.
+*
+* @param x    	  x parameter
```
```suggestion
* @param x    	  input value
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/src/main.c:
```diff
@@ -0,0 +1,56 @@
+* @param alpha    shape parameter
+* @param s        scale parameter
+* @param m        location parameter
+* @return         logpdf
```
```suggestion
* @return         evaluated logPDF
```

Let's follow the same comment structure as in `main.js`

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/src/main.c:
```diff
@@ -0,0 +1,56 @@
+		return STDLIB_CONSTANT_FLOAT64_NINF;
+	}
+	z = ( x - m ) / s;
+	return stdlib_base_ln( alpha/s ) - ( 1.0+alpha ) * stdlib_base_ln( z ) - stdlib_base_pow( z, -alpha );
```
```suggestion
	return stdlib_base_ln( alpha/s ) - ( ( 1.0+alpha ) * stdlib_base_ln( z ) ) - stdlib_base_pow( z, -alpha );
```

anandkaranubc replied:
Consistent with `main.js`

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/test/test.native.js:
This looks good.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/lib/native.js:
```diff
@@ -0,0 +1,80 @@
+/**
+* Evaluates the logarithm of the probability density function (PDF) for a Fréchet distribution with shape `alpha`, scale `s`, and location `m` at a value `x`.
+*
+* @param {number} x - input value
```
```suggestion
*
* @private
* @param {number} x - input value
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/include/stdlib/stats/base/dists/frechet/logpdf.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Returns the logpdf for a Fréchet distribution with shape `alpha`, scale `s`, and location `m` at a value `x`.
```
```suggestion
* Evaluates the logarithm of the probability density function (PDF) for a Fréchet distribution with shape `alpha`, scale `s`, and location `m` at a value `x`.
```

anandkaranubc replied:
@iraandrushko This change is still needed.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/examples/c/example.c:
```diff
@@ -0,0 +1,44 @@
+int main( void ) {
+	double x;
+	double alpha;
+	double s;
```
```suggestion
	double alpha;
	double x;
	double s;
```

Variables are ordered according to the decreasing order of their length.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/examples/c/example.c:
```diff
@@ -0,0 +1,44 @@
+		s = random_uniform( 0.0, 10.0 );
+		m = random_uniform( 0.0, 10.0 );
+		y = stdlib_base_dists_frechet_logpdf( x, alpha, s, m );
+		printf( "x: %.4lf, α: %.4lf, s: %.4lf, m: %.4lf, ln(f(x;α,s,m)): %.4lf\n", x, alpha, s, m, y );
```
```suggestion
		printf( "x: %lf, α: %lf, s: %lf, m: %lf, ln(f(x;α,s,m)): %lf\n", x, alpha, s, m, y );
```

anandkaranubc replied:
This change will also be reflected in the README file

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,145 @@
+	double alpha[ 100 ];
+	double s[ 100 ];
+	double m[ 100 ];
+	double y;
```
```suggestion
	double alpha[ 100 ];
	double m[ 100 ];
	double x[ 100 ];
	double s[ 100 ];
	double elapsed;
	double y;
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,145 @@
+		x[ i ] = random_uniform( -50.0, 50.0 );
+		alpha[ i ] = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
+		s[ i ] = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
+		m[ i ] = random_uniform( -20.0, 20.0 );
```
```suggestion
		m[ i ] = random_uniform( -20.0, 40.0 );
```

Let's keep the same range as in `benchmark.js`

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/README.md:
```diff
@@ -170,6 +170,110 @@ for ( i = 0; i < 100; i++ ) {
+        s = random_uniform( 0.0, 10.0 );
+        m = random_uniform( 0.0, 10.0 );
+        y = stdlib_base_dists_frechet_logpdf( x, alpha, s, m );
+        printf( "x: %.4lf, α: %.4lf, s: %.4lf, m: %.4lf, ln(f(x;α,s,m)): %.4lf\n", x, alpha, s, m, y );
```
This will be changed according to the above comment.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/README.md:
```diff
@@ -170,6 +170,110 @@ for ( i = 0; i < 100; i++ ) {
+int main( void ) {
+    double x;
+    double alpha;
+    double s;
```
This will be changed according to the above comment.

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/README.md:
```diff
@@ -170,6 +170,110 @@ for ( i = 0; i < 100; i++ ) {
+
+#### stdlib_base_dists_frechet_logpdf( x, alpha, s, m )
+
+Returns the the logarithm of probability density function for a Fréchet distribution with shape `alpha`, scale `s`, and location `m` at a value `x`.
```
```suggestion
Evaluates the logarithm of the [probability density function][pdf] (PDF) for a [Fréchet][frechet-distribution] distribution with shape `alpha`, scale `s`, and location `m` at a value `x`.
```

anandkaranubc commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/manifest.json:
```diff
@@ -0,0 +1,90 @@
+        "@stdlib/math/base/special/ln",
+        "@stdlib/constants/float64/ninf",
+        "@stdlib/constants/float64/nan",
+        "@stdlib/constants/float64/eps"
```
You don't need `"@stdlib/constants/float64/eps"` since you are not using `EPS` in `example.c`

iraandrushko committed:
feat: add C implementation for stats/base/dists/frechet/logpdf: fix CodeReview comments

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

iraandrushko committed:
feat: add C implementation for stats/base/dists/frechet/logpdf: fix PR comments part 2

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

iraandrushko committed:
feat: add C implementation for stats/base/dists/frechet/logpdf: fix PR comments part 3

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

iraandrushko commented:
Hi @anandkaranubc , thanks for the review comments, hope I handled them correctly

iraandrushko committed:
feat: add C implementation for stats/base/dists/frechet/logpdf: fix PR comments part 4

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/README.md:
````diff
@@ -170,6 +170,110 @@ for ( i = 0; i < 100; i++ ) {
+
+### Examples
+
+```c
+
+#include "stdlib/stats/base/dists/frechet/logpdf.h"
+#include <stdlib.h>
+#include <stdio.h>
````
````suggestion
```c
#include "stdlib/stats/base/dists/frechet/logpdf.h"
#include <stdlib.h>
#include <stdio.h>
````

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/frechet/logpdf/src/main.c:
```diff
@@ -0,0 +1,56 @@
+/**
+* Evaluates the logarithm of the probability density function (PDF) for a Fréchet distribution with shape `alpha`, scale `s`, and location `m` at a value `x`.
+*
+* @param x    	  input value
+* @param alpha    shape parameter
+* @param s        scale parameter
+* @param m        location parameter
+* @return         evaluated logPDF
```
```suggestion
* @param x       input value
* @param alpha    shape parameter
* @param s        scale parameter
* @param m        location parameter
* @return         evaluated logPDF
```

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (approved):
Thank you @iraandrushko for your contribution and @anandkaranubc for helping getting it over the finish line!
